File: src/test/java/com/polygloat/controllers/ImportControllerTest.java
package com.polygloat.controllers;

import com.github.javafaker.Faker;
import com.polygloat.dtos.PathDTO;
import com.polygloat.dtos.request.ImportDto;
import com.polygloat.model.Repository;
import lombok.SneakyThrows;
import org.testng.annotations.Test;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import static com.polygloat.Assertions.Assertions.assertThat;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

public class ImportControllerTest extends SignedInControllerTest {

    private Faker faker = new Faker();

    @SneakyThrows
    @Test
    public void testDoImport() {
        Repository base = dbPopulator.createBase(generateUniqueString());

        HashMap<String, String> data = new HashMap<>();

        for (int i = 0; i < 1000; i++) {
            String sentence = faker.lorem().sentence();
            logger.info(String.format("Creating sentence: %s", sentence));
            String source = sentence.substring(0).replace(" ", "_").replaceAll("[^A-Za-z0-9]", "").toLowerCase();
            data.put(source, sentence);
        }

        ImportDto dto = ImportDto.builder().languageAbbreviation("en").data(data).build();
        performPost("/api/repository/" + base.getId() + "/import", dto).andExpect(status().isOk());

        for (Map.Entry<String, String> entry : data.entrySet()) {
            Map<String, String> translations = translationService.getSourceTranslationsResult(base.getId(), PathDTO.fromFullPath(entry.getKey()), Set.of("en"));
            assertThat(translations.get("en")).isEqualTo(entry.getValue());
        }

    }
}

File: src/test/kotlin/io/tolgee/controllers/ImportControllerTest.kt
package io.tolgee.controllers

import com.github.javafaker.Faker
import io.tolgee.assertions.Assertions.assertThat
import io.tolgee.dtos.ImportDto
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.asyncDispatch
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.testng.annotations.Test
import java.util.*
import java.util.stream.Collectors
import kotlin.math.ceil
import kotlin.system.measureTimeMillis

class ImportControllerTest : SignedInControllerTest() {
    private val faker = Faker()

    companion object {
        const val IMPORT_SINGLE_TIME_LIMIT_MS: Double = 2.5
        fun getMaxTimeInMs(count: Int): Long = ceil(count.toDouble() * IMPORT_SINGLE_TIME_LIMIT_MS).toLong()
    }

    @Test
    fun basicImport() {
        val repository = dbPopulator.createBase(generateUniqueString())

        commitTransaction()

        val dto = generateData(5000)

        performPost("/api/repository/${repository.id}/import", dto)
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn().asyncResult

        val language = languageService.findByAbbreviation("en", repository).orElseGet(null)!!
        val translations = translationService.getAllByLanguageId(language.id)
        assertThat(translations.stream().map { it.text }.collect(Collectors.toList())).containsAll(dto.data!!.values)
    }

    @Test
    fun streamingProgress() {
        val repository = dbPopulator.createBase(generateUniqueString())

        commitTransaction()

        val dto = generateData(50)

        val asyncResult = performPost("/api/repository/${repository.id}/import", dto)
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val result = mvc.perform(asyncDispatch(asyncResult)).andReturn()
        assertThat(result.response.contentAsByteArray).startsWith(0, 1, 2).endsWith(dto.data!!.size - 1)
    }

    @Test
    fun basicPerformance() {
        val repository = dbPopulator.createBase(generateUniqueString())

        commitTransaction()

        val dto1 = generateData(10000)

        val timeOnEmpty = measureTimeMillis {
            performPost("/api/repository/${repository.id}/import", dto1)
                    .andExpect(MockMvcResultMatchers.status().isOk)
                    .andReturn().asyncResult
        }

        assertThat(timeOnEmpty).isLessThan(getMaxTimeInMs(dto1.data!!.size))

        val dto2 = generateData(5000)
        val timeOnFull = measureTimeMillis {
            performPost("/api/repository/${repository.id}/import", dto2)
                    .andExpect(MockMvcResultMatchers.status().isOk)
                    .andReturn().asyncResult
        }

        //if it does not import in 5 seconds, you are doing something terribly wrong!
        assertThat(timeOnFull).isLessThan(getMaxTimeInMs(dto2.data!!.size))
    }

    @Test
    fun sameData() {
        val repository = dbPopulator.createBase(generateUniqueString())

        commitTransaction()

        val dto = generateData(100)

        performPost("/api/repository/${repository.id}/import", dto)
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn().asyncResult

        val language = languageService.findByAbbreviation("en", repository).orElseGet(null)!!
        var translations = translationService.getAllByLanguageId(language.id)
        assertThat(translations.stream().map { it.text }.collect(Collectors.toList())).containsAll(dto.data!!.values)

        performPost("/api/repository/${repository.id}/import", dto)
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn().asyncResult

        translations = translationService.getAllByLanguageId(language.id)
        assertThat(translations.stream().map { it.text }.collect(Collectors.toList())).containsAll(dto.data!!.values)
    }

    private fun generateData(count: Int): ImportDto {
        val data = HashMap<String, String>()
        for (i in 1..count) {
            val sentence = faker.lorem().sentence()
            val key = sentence.replace(" ", "_")
                    .replace("[^A-Za-z0-9]".toRegex(), "")
                    .toLowerCase()
            data[key + i] = sentence
        }

        return ImportDto("en", data)
    }
}